blog: Add GSoC'24 post on testing mimalloc memory allocator (#443)
unikraft/docs: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx

This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to optimize the performance of their applications.

### Objectives

[Hugo Lefeuvre](https://github.com/hlef) has made an effort to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
However, as Unikraft has evolved significantly over the years, more work is needed to adapt the `lib-mimalloc` port to the latest Unikraft core.

The steps of porting the memory allocators are as follows:

1. Adapt the existing port, which uses `lib-newlib` and `lib-pthread-embedded` as the C library to use `lib-musl` instead
1. Patch the existing port of mimalloc (v1.6.3) to adapt to the latest Unikraft's memory allocation interface
1. Patch the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0)

## Current Progress

Last month, I successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc v1.6

> .3

Reviewer: Put on previous line

Author: The Linter has a "one end of sentence punctuation per line" rule so it doesn't want to see two `.`s in one line... Do you know of any way to override that?

Author: Nvm - I put them in a code block and it works just fine!

> against the latest Unikraft core (v2.1.7) with `lib-musl`.

Reviewer: ```suggestion
against the latest Unikraft core (v2.1.7) with `lib-musl`.
```

Should it be v0.17.0?

Author: Yes it should 👍

> Following that, this month I am focused on stress-testing the memory allocator to validate its usability and correctness.

Reviewer: ```suggestion
Following that, this month I am focused on stress-testing the `mimalloc` memory allocator to validate its usability and correctness.
```

### Benchmark

> Right after porting, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.

Reviewer: ```suggestion
After porting was completed, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.
```

> That having no problem, I ported `cache-scratch`, a multi-threaded memory allocation benchmark, to Unikraft.

Reviewer: Make a new paragraph with this

It creates a number of worker threads that allocate a given-sized object, repeatedly write on it, then free it.
An allocator that allows multiple threads to re-use the same small object (possibly all in one cache-line) will scale poorly, while an allocator like mimalloc, which gives each thread a private and page-aligned heap, will exhibit near-linear performance scaling.

### Porting `cache-scratch`

It is fairly intuitive to port the C program itself to Unikraft as the Musl library already provides most of what we need, including the `pthread` interface.
It would be ideal if we could separate the helper functions from the main logic and build the benchmark as a library (as we do need to test the memory allocator against more benchmarks in the future), because then we could just build an all-in-one image and run a large number of tests with different benchmarks and different parameters using an automated script.
However, I haven't figured out a good way to do it as [Unikraft's build system](https://unikraft.org/docs/internals/build-system#makefileuk) is quite complex.

> So for now, the benchmark is ported as a single `main.c` file that will run as an application.

Reviewer: Do you have some preliminary results to put into the blogpost?

Author: I do have the code ready - where should I put them?

Reviewer 2: Anywhere. You can create a repo on your github, or a gist or anything, lust add a link here.

Author: Done

I also added tips about turning off compiler optimization for debugging [here](https://github.com/unikraft/docs/pull/442).

## Next steps

First, I will continue working on the `cache-scratch` benchmark:

1. Debug `cache-scratch`: Figure out what went wrong in the multi-threaded test
1. Run `cache-scratch` with different sets of arguments to validate the memory allocator under different conditions

After that, I will test mimalloc with other more complicated benchmarks, which may include:

- `threadtest`: tests general scalability of the allocator;
Each thread repeatedly allocates a number of number of objects, does a fixed amount of computation, and then deletes its objects.
- `linux-scalability`: a basic scalability test from Linux libc malloc testing, similar to `threadtest` except that there is no extra 'work' between allocations and frees so contention may be higher.
- `cache-thrash`: tests for active false sharing
- `larson`: simulates a server by having each thread allocate and deallocate objects and then transfer some objects to other threads to be freed
- `phong` - a benchmark that randomly selects allocation sizes, and randomly chooses an allocated item to free

> Once fully tested, I will continue on with the work items outlined in [my previous post].(https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft#next-steps)

Reviewer 2: ```suggestion
Once fully tested, I will continue on with the work items outlined in [my previous post](https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft#next-steps).
```

Meanwhlie, I will also try to debug the `test_cmp_requeue_two_waiters_no_requeue` to the best of my ability.

## Acknowledgements

Special thanks to Răzvan Vîrtan and Radu Nichita, my two amazing mentors, for their support along the way.

> I would also like to thank Razvan Deaconescu, Hugo Lefeuvre, Ștefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.

Reviewer: ```suggestion
I would also like to thank Răzvan Deaconescu, Hugo Lefeuvre, Ștefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.
```
